x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/upgrade_prebuilt_rules_table/upgrade_prebuilt_rules_table_context.tsx

```diff
@@ -168,92 +169,142 @@ export const UpgradePrebuiltRulesTableContextProvider = ({
-        .map((ruleId) => ({
-          rule_id: ruleId,
-          version:
-            rulesUpgradeState[ruleId].diff.fields.version?.target_version ??
-            rulesUpgradeState[ruleId].current_rule.version,
-          revision: rulesUpgradeState[ruleId].revision,
-          fields: isPrebuiltRulesCustomizationEnabled
-            ? constructRuleFieldsToUpgrade(rulesUpgradeState[ruleId])
-            : undefined,
-        }));
-
-      setLoadingRules((prev) => [...prev, ...ruleUpgradeSpecifiers.map((x) => x.rule_id)]);
+      const upgradingRuleIds = ruleIds.filter((ruleId) => !conflictRuleIdsSet.has(ruleId));
+      const ruleUpgradeSpecifiers: RuleUpgradeSpecifier[] = upgradingRuleIds.map((ruleId) => ({
+        rule_id: ruleId,
+        version: rulesUpgradeState[ruleId].target_rule.version,
+        revision: rulesUpgradeState[ruleId].revision,
+        fields: constructRuleFieldsToUpgrade(rulesUpgradeState[ruleId]),
+      }));
+
+      setLoadingRules((prev) => [...prev, ...upgradingRuleIds]);
 
       try {
         // Handle MLJobs modal
-        if (shouldConfirmMLJobs && !(await confirmLegacyMLJobs())) {
+        if (!(await confirmLegacyMLJobs())) {
           return;
         }
 
-        if (
-          isPrebuiltRulesCustomizationEnabled &&
-          conflictRuleIdsSet.size > 0 &&
-          !(await confirmConflictsUpgrade())
-        ) {
+        if (conflictRuleIdsSet.size > 0 && !(await confirmConflictsUpgrade())) {
           return;
         }
 
-        await upgradeSpecificRulesRequest(ruleUpgradeSpecifiers);
+        await upgradeSpecificRulesRequest({
+          pickVersion: 'MERGED',
+          rules: ruleUpgradeSpecifiers,
+        });
       } catch (err) {
         addError(err, { title: i18n.UPDATE_ERROR });
       } finally {
-        const upgradedRuleIdsSet = new Set(ruleUpgradeSpecifiers.map((x) => x.rule_id));
+        const upgradedRuleIdsSet = new Set(upgradingRuleIds);
 
         setLoadingRules((prev) => prev.filter((id) => !upgradedRuleIdsSet.has(id)));
       }
     },
     [
       confirmLegacyMLJobs,
       confirmConflictsUpgrade,
-      shouldConfirmMLJobs,
       rulesUpgradeState,
       upgradeSpecificRulesRequest,
-      isPrebuiltRulesCustomizationEnabled,
       addError,
     ]
   );
 
-  const ruleActionsFactory = useCallback(
-    (rule: RuleResponse, closeRulePreview: () => void) => (
-      <EuiButton
-        disabled={
-          loadingRules.includes(rule.rule_id) ||
-          isRefetching ||
-          isUpgradingSecurityPackages ||
-          rulesUpgradeState[rule.rule_id]?.hasUnresolvedConflicts
+  const upgradeRulesToTarget = useCallback(
+    async (ruleIds: RuleSignatureId[]) => {
+      const ruleUpgradeSpecifiers: RuleUpgradeSpecifier[] = ruleIds.map((ruleId) => ({
+        rule_id: ruleId,
+        version:
+          rulesUpgradeState[ruleId].diff.fields.version?.target_version ??
+          rulesUpgradeState[ruleId].current_rule.version,
```

Comment:
Using `diff.fields.version?.target_version` or `rulesUpgradeState[ruleId].current_rule.version` looks weird or I miss something here.

The logic is straightforward. There is a `diff` if we have some fields changed in a new prebuilt rules package  version. When one or more fields changed it's expected to have `version` bumped. This way `version` will always be in the `diff` and it equals to `target_rule.version`.

Could we use `version: rulesUpgradeState[ruleId].target_rule.version` here?

Reply:
I don't know why this logic was written in this specific way, but using `version: rulesUpgradeState[ruleId].target_rule.version` seems more straightforward to me